lib/improver/convection.py

```diff
@@ -0,0 +1,207 @@
+        self.unweighted_mode = unweighted_mode
+        self.ens_factor = ens_factor
+        self.use_adjacent_grid_square_differences = (
+            use_adjacent_grid_square_differences)
+
+    def __repr__(self):
+        """Represent the configured plugin instance as a string."""
+        result = ('<DiagnoseConvectivePrecipitation: lower_threshold {}; '
+                  'higher_threshold {}; neighbourhood_method: {}; '
+                  'radii_in_km: {}; fuzzy_factor {}; '
+                  'below_thresh_ok: {}; lead_times: {}; '
+                  'unweighted_mode: {}; ens_factor: {}; '
+                  'use_adjacent_grid_square_differences: {}>')
+        return result.format(
+            self.lower_threshold, self.higher_threshold,
+            self.neighbourhood_method, self.radii_in_km, self.fuzzy_factor,
+            self.below_thresh_ok, self.lead_times, self.unweighted_mode,
+            self.ens_factor, self.use_adjacent_grid_square_differences)
+
+    def _calculate_convective_ratio(self, cube):
+        """
+        Calculate the convective ratio by:
+        1. Threshold the input cube using the lower_threshold and
+           apply neighbourhood processing.
+        2. Threshold the input cube using the higher_threshold and
+           apply neighbourhood processing.
+        3. Calculate the convective ratio by:
+           higher_threshold_cube / lower_threshold_cube.
+           For example, the higher_threshold might be 5 mm/hr, whilst the
+           lower_threshold might be 0 mm/hr.
+
+        The Convective Ratio can have the following values:
+            * A non-zero fractional value, indicating that both the higher
+              and lower thresholds were exceeded.
+            * A zero value, if the lower threshold was exceeded, whilst the
+              higher threshold was not exceeded.
+            * A NaN value (np.nan), if neither the higher or lower thresholds
+              were exceeded, and therefore a NaN value results from the
+              0/0 division.
+
+        Args:
+            cube : Iris.cube.Cube
+                The cube from which the convective ratio will be calculated.
+
+        Returns:
+            convective_ratio : Iris.cube.Cube
+                Cube containing the convective ratio
+        """
+        threshold_dict = {}
+
+        for threshold in [self.lower_threshold, self.higher_threshold]:
+            threshold_cube = (
+                BasicThreshold(
+                    threshold, fuzzy_factor=self.fuzzy_factor,
+                    below_thresh_ok=self.below_thresh_ok).process(cube.copy()))
+            neighbourhooded_cube = NeighbourhoodProcessing(
+                self.neighbourhood_method, self.radii_in_km,
+                lead_times=self.lead_times,
+                unweighted_mode=self.unweighted_mode,
+                ens_factor=self.ens_factor).process(threshold_cube)
+            threshold_dict[threshold] = neighbourhooded_cube
+
```

Comment:
Actually following discussion with @fionaRust, I'd misunderstood the relevance of the first acceptance criterion, so I'm making corrections to deal with that at the moment.